Update rust reference info about closures (#1)
sexxi-goose/reference: src/types/closure.md

<!-- ignore: continuation of above -->
```rust,ignore
f(Closure{ left_top: rect.left_top, right_bottom_x: rect.left_top.x });
```

## Capture modes

> The compiler prefers to capture a value basedimmutable borrow,

Author: ```suggestion
The compiler prefers to capture a value by immutable borrow,
```

followed by unique immutable borrow (see below), by mutable borrow, and finally
by value (move or copy). It will pick the first choice of these that allows the closure to
compile. The choice is made only with regards to the contents of the closure
expression; the compiler does not take into account surrounding code, such as
the lifetimes of involved variables or fields.

## Capture Precision

> The precise path that gets captured is typically the full path that is used in the closure, but there are cases where we will only capture a prefix of the path.

Author: ```suggestion

The precise path that gets captured is typically the full path that is used in the closure, but there are cases where we will only capture a prefix of the path.
```

### Shared prefix

> In the case where a path and one of the ancestor’s of that path are both captured by a closure, the ancestor path is captured with `CaptureMode = max(AncestorCaptureMode, DescendantCaptureMode)` using the strict weak ordering 

Author: ```suggestion
In the case where a path and one of the ancestor’s of that path are both captured by a closure, the ancestor path is captured with the highest capture mode among the two captures,`CaptureMode = max(AncestorCaptureMode, DescendantCaptureMode)`, using the strict weak ordering 
```

`ImmBorrow < UniqueImmBorrow < MutBorrow < ByValue`.

Note that this might need to be applied recursively.

```rust=
let s = String::new("S");
let t = (s, String::new("T"));
let mut u = (t, String::new("U"));

let c = || {
    println!("{:?}", u); // u captured by ImmBorrow
    u.0.truncate(0); // u.0 captured by MutBorrow
    move_value(u.0.0); // u.0.0 captured by ByValue
};
```

Overall the closure will capture `u` by `ByValue`.

### Wild Card Patterns
Closures only capture data that needs to be read, which means the following closures will not capture `x`

```rust
let x = 10;
let c = || {
    let _ = x;
};

let c = || match x {
    _ => println!("Hello World!")
};
```

### Capturing references in move contexts

> Rust doesn't allow moving fields out references as a result in the case of move closures, when the closure takes ownership of a path, it will only take ownership of a prefix of path that runs up until the first dereference (this may be the entire path if there are no dereferences in the path).

Reviewer: moving fields out of

Reviewer: Instead of truncate ... we reborrow the data being accessed throw the reference

> Rust doesn't allow moving fields out references as a result in the case of move closures, when the closure takes ownership of a path, it will only take ownership of a prefix of path that runs up until the first dereference (this may be the entire path if there are no dereferences in the path).

Author: ```suggestion
Rust doesn't allow moving fields out of references. As a result, in the case of move closures, when a closure takes ownership of a path, it will only take ownership of a prefix of the path that runs up until the first dereference (this may be the entire path if there are no dereferences in the path).
```

```rust
struct T(String, String);

let mut t = T(String::from("foo"), String::from("bar"));
let t = &mut t;
let c = move || t.0.truncate(0); // closure captures (&mut t.0)
```

### Raw pointer dereference

> In Rust it's `unsafe` to dereference a raw pointer therefore closures will only capture the prefix of a path that runs upto but not including the first dereference of a raw pointer.

Author: ```suggestion
In Rust, it's `unsafe` to dereference a raw pointer. Therefore, closures will only capture the prefix of a path that runs up to, but not including, the first dereference of a raw pointer.
```

```rust,
struct T(String, String);

> let foo = T(String::from("foo"), String::from("bar"));

Author: ```suggestion
let t = T(String::from("foo"), String::from("bar"));
```

> let foo = &foo as *const T;

Author: ```suggestion
let t = &t as *const T;
```

let c = || unsafe {

>     println!("{}", (*foo).0); // closure captures t

Author: ```suggestion
    println!("{}", (*t).0); // closure captures t
```

};
```

### Reference into unaligned `struct`s

> In Rust it's `unsafe` to hold references to unaligned fields in a structure and therefore closures will only capture the prefix of the path that runs upto but not including the first field access into an unaligned struct.

Author: ```suggestion
In Rust, it's `unsafe` to hold references to unaligned fields in a structure, and therefore, closures will only capture the prefix of the path that runs up to, but not including, the first field access into an unaligned structure.
```

```rust
#[repr(packed)]
struct T(String, String);

let t = T(String::from("foo"), String::from("bar"));
let c = || unsafe {
    println!("{}", t.0); // closure captures t
};
```

### `Box` vs other `Deref` implementations

> The compiler special cases the implementation of the Deref trait for `Box` and the feature treats `Box` as a special entity as well.

Author: ```suggestion
The compiler treats the implementation of the Deref trait for `Box` differently, as it is considered a special entity.
```

> To draw a comparison we look at explamples involving `Rc` and `Box`. The `*rc` is desugared to a call to the trait method `deref` method defined on `Rc`, but `*box` is special cased by the compiler. This implies that the compiler is able to do precise capture on contents of the Box.

Author: ```suggestion
For example, let us look at examples involving `Rc` and `Box`. The `*rc` is desugared to a call to the trait method `deref` defined on `Rc`, but since `*box` is treated differently by the compiler, the compiler is able to do precise capture on contents of the `Box`.
```

#### Non `move` closure

> In a closure that isn't `move`, if the contents of the box are not moved into the closure body the contents of the box are precisely captured.

Author: ```suggestion
In a non `move` closure, if the contents of the `Box` are not moved into the closure body, the contents of the `Box` are precisely captured.
```

```rust
# use std::rc::Rc;

struct S(i32);

let b = Box::new(S(10));
let c_box = || {
    println!("{}", (*b).0); // closure captures `(*b).0`
};
    
let r = Rc::new(S(10));
let c_rc = || {
    println!("{}", (*r).0); // closure caprures `r`
};
```

> However if the contents of the box are moved into the closure then the box is entierly captured. This is done to do so the amount of data that needs to be moved into the closure is minimized.

Author: ```suggestion
However, if the contents of the `Box` are moved into the closure, then the box is entirely captured. This is done so the amount of data that needs to be moved into the closure is minimized.
```

```rust
struct S(i32);

let b = Box::new(S(10));
let c_box = || {
    let x = (*b).0; // closure captures `b`
};
```

#### `move` closure

> Similar to moving contents of a box in a non-`move` closure reading the contents of a box in a `move` closure results in capturing the box entierly.

Author: ```suggestion
Similarly to moving contents of a `Box` in a non-`move` closure, reading the contents of a `Box` in a `move` closure will capture the `Box` entirely.
```

[`Sync`]: ../special-types-and-traits.md#sync
[closure expression]: ../expressions/closure-expr.md
[derived]: ../attributes/derive.md

## Drop Order

> If a closure captures a field of a composite types such as structs, tuples, and enums by ByValue, the field's lifetime would now be tied to the closure. As a result, it is possible for disjoint fields of a composite types to be dropped at different times.

Author: ```suggestion
If a closure captures a field of a composite types such as structs, tuples, and enums by value, the field's lifetime would now be tied to the closure. As a result, it is possible for disjoint fields of a composite types to be dropped at different times.
```

```rust
{
    let tuple = 
      (String::from("foo"), String::from("bar")); // --+
    { //                                               |
        let c = || { // ----------------------------+  |
            // tuple.0 is captured into the closure |  |
            drop(tuple.0); //                 |  |
        }; //                                       |  |
    } // 'c' and 'tuple.0' dropped here ------------+  |
} // tuple.1 dropped here -----------------------------+
```

> ## Wild Card Patterns

Reviewer: this section is already part of the Capture precision above

> Closures now only capture data that needs to be read, which means the following closures will not capture `x`

Author: ```suggestion
Closures only capture data that needs to be read, which means the following closures will not capture `x`
```

```rust
let x = 10;
let c = || {
    let _ = x;
};

let c = || match x {
    _ => println!("Hello World!")
};
```

# Edition 2018 and before

## Closure types difference

> In Edition 2018 and before, a closure would capture variables in its entirety. This meant that for the example used in the [Closure types]() section the generated closure type would look something like this:

Author: ```suggestion
In Edition 2018 and before, a closure would capture variables in its entirety. This means that for the example used in the [Closure types] section, the generated closure type would instead look something like this:
```

Reviewer: we should create a link above and link them here? I think if look at a rust RFC there should be expamples for it.

I'm refering to the fact that [Clsoure types] doesn't have a link associated with it

<!-- ignore: simplified -->
```rust,ignore
struct Closure<'a> {
    rect : &'a mut Rectangle,
}

impl<'a> FnOnce<()> for Closure<'a> {
    type Output = String;
    fn call_once(self) -> String {        
        self.rect.left_top.x += 1;
        self.rect.right_bottom.x += 1;
        format!("{:?}", self.rect.left_top)
    }
}
```

> and the call to `f` would work as if it were:

Author: ```suggestion
and the call to `f` would work as follows:
```

<!-- ignore: continuation of above -->
```rust,ignore
f(Closure { rect: rect });
```

## Capture precision difference

Composite types such as structs, tuples, and enums are always captured in its intirety,

> not by individual fields. It may be necessary to borrow into a local variable in order to capture a single field:

Author: ```suggestion
not by individual fields. As a result, it may be necessary to borrow into a local variable in order to capture a single field:
```

```rust
# use std::collections::HashSet;
#
struct SetVec {
    set: HashSet<u32>,
    vec: Vec<u32>
}

impl SetVec {
    fn populate(&mut self) {
        let vec = &mut self.vec;
        self.set.iter().for_each(|&n| {
            vec.push(n);
        })
    }
}
```

If, instead, the closure were to use `self.vec` directly, then it would attempt
to capture `self` by mutable reference. But since `self.set` is already
borrowed to iterate over, the code would not compile.

If the `move` keyword is used, then all captures are by move or, for `Copy`
types, by copy, regardless of whether a borrow would work. The `move` keyword is
usually used to allow the closure to outlive the captured values, such as if the
closure is being returned or used to spawn a new thread.

Regardless of if the data will be read by the closure, i.e. in case of wild card patterns, if a variable defined outside the closure is mentioned within the closure the variable will be captured in its entirety.

> ## Other traits difference

Author: I think we can actually delete this section. This is because the modified rules stated "the captured values" and we already clarified in the earlier Edition2018 subsection that the values captured are variables that are captured in its entirety

Reviewer: I'm fine removing it

In Edition 2018 and before, the rules outlined in the [Other traits]() section would need to apply instead for entire variables. That is for a closure to implement a specific trait, all the captured variables must implement the specific trait.

## Drop order difference

> As composite types are now fully captured, a closure which captures one of those composite types ByValue would drop it as the same time as closure gets dropped.

Author: ```suggestion
As composite types are captured in their entirety, a closure which captures one of those composite types by value would drop the captured variable at the same time as closure gets dropped.
```

Reviewer: would drop the entire captured variable
